torchaudio/datasets/utils.py

```diff
@@ -76,41 +65,161 @@ def makedir_exist_ok(dirpath):
+    # Detect filename
+    filename = filename or req_info.get_filename() or os.path.basename(url)
+    filepath = os.path.join(download_folder, filename)
+
+    if resume and os.path.exists(filepath):
+        mode = "ab"
+        local_size = os.path.getsize(filepath)
+    elif resume and os.path.exists(filepath):
+        raise RuntimeError(
+            "{} already exists. Delete the file manually and retry.".format(filepath)
+        )
     else:
-        try:
-            print("Downloading " + url + " to " + fpath)
-            urllib.request.urlretrieve(url, fpath, reporthook=gen_bar_updater())
-        except (urllib.error.URLError, IOError) as e:
-            if url[:5] == "https":
-                url = url.replace("https:", "http:")
-                print(
-                    "Failed download. Trying https -> http instead."
-                    " Downloading " + url + " to " + fpath
-                )
-                urllib.request.urlretrieve(url, fpath, reporthook=gen_bar_updater())
-            else:
-                raise e
+        mode = "wb"
+        local_size = None
+
+    if hash_value and local_size == int(req_info.get("Content-Length", -1)):
+        if validate_file(filepath, hash_value, hash_type):
+            return
+        raise RuntimeError(
+            "The hash of {} does not match. Delete the file manually and retry.".format(
+                filepath
+            )
+        )
+
+    with open(filepath, mode) as fpointer:
+        for chunk in stream_url(url, start_byte=local_size, progress_bar=progress_bar):
+            fpointer.write(chunk)
+
+    if hash_value and not validate_file(filepath, hash_value, hash_type):
+        raise RuntimeError(
+            "The hash of {} does not match. Delete the file manually and retry.".format(
+                filepath
+            )
+        )
+
+
+def download_url(urls, *args, max_workers=5, **kwargs):
+    """Download urls to disk. The other arguments are passed to download_single_url.
+
+    Args:
+        urls (str or List[str]): List of urls.
+        max_workers (int): Maximum number of workers.
+    """
+
+    if isinstance(urls, str):
+        return download_single_url(urls, *args, **kwargs)
+
+    # Turn arguments into lists
+    args = list(args)
+    for i, item in enumerate(args):
+        if not isinstance(item, list):
+            args[i] = [item] * len(urls)
+    args = list(zip(*args))
+
+    # Turn keyword arguments into lists
+    for key, value in kwargs.items():
+        if not isinstance(value, list):
+            kwargs[key] = [value] * len(urls)
+    kwargs = [dict(zip(kwargs.keys(), values)) for values in zip(*(kwargs.values()))]
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
+        futures = [
+            executor.submit(download_single_url, url, *arg, **kwarg)
+            for url, arg, kwarg in zip(urls, args, kwargs)
+        ]
+        return concurrent.futures.as_completed(futures)
```

Comment:
This notation and concurrent futures are not supported in python 2. We could leave the multithreaded download for the first release with python 3. Otherwise, I could add some mechanism to offer this functionality only when python 3 is available and default to single thread otherwise.